src/ripple/app/paths/RippleState.cpp

```diff
@@ -66,17 +65,17 @@ RippleState::getJson(int)
     return ret;
 }
 
-std::vector<RippleState::pointer>
+std::vector<RippleState>
 getRippleStateItems(AccountID const& accountID, ReadView const& view)
 {
-    std::vector<RippleState::pointer> items;
+    std::vector<RippleState> items;
     forEachItem(
         view,
         accountID,
         [&items, &accountID](std::shared_ptr<SLE const> const& sleCur) {
             auto ret = RippleState::makeItem(accountID, sleCur);
```

Comment:
I'd remove `makeItem` entirely; `forEachItem` will never pass in a `sleCur` that's `nullptr` and you can just inline the check for `ltRIPPLE_STATE`:

    [&items, &accountID](std::shared_ptr<SLE const> const& sleCur) {
        if (sleCur->getType() == ltRIPPLE_STATE)
            item.emplace_back(sleCur, accountID);
    }

You'll need to run this check in the other place where `makeItem` is invoked, but overall it's cleaner and simpler.

Reply:
```
You'll need to run this check in the other place where makeItem is invoked, but overall it's cleaner and simpler.
```
We should talk about this. Since there's no way for a constructor to signal an error (other than an exception, which we disallow), the our options on error are:
 1) it's a logic error - terminate the program; 
2) an `bool OK()` function that needs to be checked after construction. 
(2) Doesn't work well in practice. It's too easy to forget to call.
(1) is a reasonable option if the constructor is only called in a very limited context. Then it's easy to audit all the callers to make sure the preconditions are met, and we don't expect new calls to be added.

I think this is class is too generally useful to expect no new calls to be added. I think a `make` function is the least bad option we have.

I'm very open to discussion on this. But for now, I vote we keep the `make` function. We should also talk about relaxing our restrictions on throwing.